src/platforms/posix/drivers/bebop_flow/bebop_flow.cpp

```diff
@@ -0,0 +1,322 @@
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file bebop_flow.cpp
+ *
+ * This is a wrapper around the Parrot Bebop's downward-facing camera and integrates
+ * an optical flow computation.
+ */
+
+#include <stdint.h>
+
+#include <px4_tasks.h>
+#include <px4_getopt.h>
+#include <px4_posix.h>
+
+#include "video_device.h"
+#include "dump_pgm.h"
+#include <mt9v117/MT9V117.hpp>
+
+extern "C" { __EXPORT int bebop_flow_main(int argc, char *argv[]); }
+
+using namespace DriverFramework;
+
+namespace bebop_flow
+{
+MT9V117 *image_sensor = nullptr;				 // I2C image sensor
+VideoDevice *g_dev = nullptr;            // interface to the video device
+volatile bool _task_should_exit = false; // flag indicating if bebop flow task should exit
+static bool _is_running = false;         // flag indicating if bebop flow  app is running
+static px4_task_t _task_handle = -1;     // handle to the task main thread
+volatile unsigned int _trigger = 0;			 // Number of images to write as pgm
+
+const char *dev_name = "/dev/video0";		 // V4L video device
+
+int start();
+int stop();
+int info();
+int trigger(int count);
+int clear_errors();
+void usage();
+void task_main(int argc, char *argv[]);
+
+void task_main(int argc, char *argv[])
+{
+	_is_running = true;
+	int ret = 0;
+	struct frame_data frame;
+	memset(&frame, 0, sizeof(frame));
+	uint32_t timeout_cnt = 0;
+
+	// Main loop
+	while (!_task_should_exit) {
+
+		ret = g_dev->get_frame(frame);
+
+		if (ret < 0) {
+			PX4_ERR("Get Frame failed");
+			continue;
+
+		} else if (ret == 1) {
+			// No image in buffer
+			usleep(1000);
+			++timeout_cnt;
+
+			if (timeout_cnt > 1000) {
+				PX4_WARN("No frames received for 1 sec");
+				timeout_cnt = 0;
+			}
+
+			continue;
+		}
+
+		timeout_cnt = 0;
+
+		// Write images into a file
+		if (_trigger > 0) {
+			PX4_INFO("Trigger camera");
+
+			dump_pgm(frame.data, frame.bytes, frame.seq, frame.timestamp);
+			--_trigger;
+		}
+
+		/***************************************************************
+		 *
+		 * Optical Flow computation
```

Comment:
Is it correct that the flow code is supposed to go here instead of a callback function?